### ISSUE TITLE
Windows: Investigate custom menu implementation

### ISSUE THREAD

**Issue by bpasero** [Microsoft/vscode#29024]
To fix 35+ menu issues: https://github.com/Microsoft/vscode/issues?q=is%3Aopen+is%3Aissue+label%3Amenu

**Comment by WaiSiuKei** [comment]
Hi @bpasero, you can implement the menu in JS in a Browserwindow and then rendering the workbench in a Webview.

**Comment by bpasero** [comment]
@WaiSiuKei why so complicated, the window can be frameless and own everything, including menu and title?

**Comment by WaiSiuKei** [comment]
Yes, the browserWindow can be frameless, with menu and title created by js. The reason is that when you click the menu, the activeElement of the webview will no change, on the webview itself will be blur, then  a menu item clicked, you can focus the webview back and send the actionId to webview just like the menu works now. The fous/blur issue is most concerned. Take [ghedit](https://github.com/spiffcode/ghedit) for example, it inserts the menu and before the workbench, but you will find that there is no copy/paste in the menu, it cannot impletement the functionality because the focus of editor is lost.

**Comment by bpasero** [comment]
Ah good point, needs more thinking indeed.

**Comment by sbatten** [comment]
Results of investigation with other applications. The goal of this investigation is to see if a custom menu bar will solve the problems/issues being reported to us by testing other applications that have gone this route. GitHub Desktop use a combined menu/title and an Atom Plugin provides a similar design. We use notepad as a baseline that uses a true native menu bar to compare.

Accessibility

Issue | GitHub Desktop | Atom Plugin |	Notepad	|	Result
---|---|---|---|---:
[27106](https://github.com/Microsoft/vscode/issues/27106) | Not Fixed | Fixed | Fixed | **Fixed**
[25418](https://github.com/Microsoft/vscode/issues/25418) | Fixed	| Fixed |	Not Fixed	| **Fixed**
[19011](https://github.com/Microsoft/vscode/issues/19011) | Fixed	| Fixed	|	Fixed | **Fixed**
[17224](https://github.com/Microsoft/vscode/issues/17224) | Not Fixed | Not Fixed	| Not Fixed	| **Not Fixed**
[17180](https://github.com/Microsoft/vscode/issues/17180) | Fixed	|	Not Fixed	|	Not Fixed	| **Fixed**
[17177](https://github.com/Microsoft/vscode/issues/17177) | - | -	| - | **Fixable**
[17172](https://github.com/Microsoft/vscode/issues/17172) | Partial	| Fixed	|	Partial	| **Fixed**
[6140](https://github.com/Microsoft/vscode/issues/6140)	 | Not Fixed | Partial | Fixed | **Fixed**
[5556](https://github.com/Microsoft/vscode/issues/5556)  | - | - | - | **Dupe of 27106**
[2945](https://github.com/Microsoft/vscode/issues/2945)	 | - | - | - | **Dupe of 1569**
[1569](https://github.com/Microsoft/vscode/issues/1569)  | Fixed | Fixed |	Fixed	| **Fixed**

Other Issues

Description | Link
---|---:
Alt-Key conflicts with custom keybindings | [41070](https://github.com/Microsoft/vscode/issues/41070), [35010](https://github.com/Microsoft/vscode/issues/35010), [22488](https://github.com/Microsoft/vscode/issues/22488)
Support for Global App Menu in Gnome | [36667](https://github.com/Microsoft/vscode/issues/36667)
Support when clauses in MenuBar Actions | [35225](https://github.com/Microsoft/vscode/issues/35225)
Request for Custom Menu Bar | [34719](https://github.com/Microsoft/vscode/issues/34719), [16363](https://github.com/Microsoft/vscode/issues/16363)
Request for Custom Title Bar | [29997](https://github.com/Microsoft/vscode/issues/29997)
Chords and Shortcuts displayed differently | [33677](https://github.com/Microsoft/vscode/issues/33677)
Toggling Menu Bar causes issues | [21889](https://github.com/Microsoft/vscode/issues/21889), [16487](https://github.com/Microsoft/vscode/issues/16487), [32853](https://github.com/Microsoft/vscode/issues/32853), [10941](https://github.com/Microsoft/vscode/issues/10941)
Dynamic Menu Bar Contributions | [21131](https://github.com/Microsoft/vscode/issues/21131), [22359](https://github.com/Microsoft/vscode/issues/22359), [16627](https://github.com/Microsoft/vscode/issues/16627)
Localization | [4771](https://github.com/Microsoft/vscode/issues/4771)
Editor context-aware menus | [991](https://github.com/Microsoft/vscode/issues/991)

**Comment by kizomanizo** [comment]
When will this issue get some love from the DEV's? it will really be easy to the eyes.

**Comment by PurpleGranite** [comment]
Good to see this in the June 2018 Iteration plan - Item 33677 listed above should however be listed as an Accessibility issue due to the impact that differential positioning, shading and screen reader accessibility caused by that item for sight-impaired developers.

**Comment by sbatten** [comment]
@eamodio yes, as of this moment they are just light and dark, but theming is coming.

**Comment by MulverineX** [comment]
How do we use this

**Comment by sbatten** [comment]
@MulverineX Thank you for your interest. It is not currently enabled by default, but you can enable on Insiders by setting `window.titleBarStyle` to `custom`

**Comment by sbatten** [comment]
This is enabled by default for Windows in Insiders and we intend to make it default for the stable release this milestone.
